### DISCUSSION TITLE
Example and default values with dedicated named type enum

### DISCUSSION THREAD

**Discussion by pboguslawski** [swaggest/rest#132] [Q&A] [answered]
According to https://pkg.go.dev/github.com/swaggest/jsonschema-go#Enum one can use (1) dedicated named type or (2) field tag to set up enum.

With (2) it's possible do define default value for openapi.json; with
```
SortDir string   `query:"sort_dir" enum:"asc,desc" default:"asc" description:"Sort direction." example:"desc"`
```
one gets
```
{
      "name": "sort_dir",
      "in": "query",
      "description": "Sort direction.",
      "schema": {
       "enum": [
        "asc",
        "desc"
       ],
       "type": "string",
       "description": "Sort direction.",
       "default": "asc",
       "example": "desc"
      }
     },
```

Is it possibe get `example` and `default` specified in openapi.json for enum defined using (1)? With

```
type SortDir string

func (SortDir) Enum() []interface{} {
	return []interface{}{
		"asc",
		"desc",
	}
}
[...]

SortDir SortDir  `query:"sort_dir" default:"asc" description:"Sort direction." example:"desc"`
```
one gets stuff without `example` and `default` specified:
```
{
      "name": "sort_dir",
      "in": "query",
      "description": "Sort direction.",
      "schema": {
       "$ref": "#/components/schemas/SortDir"
      }
     },
[...]
"SortDir": {
    "enum": [
     "asc",
     "desc"
    ],
    "type": "string"
   }
```

**Comment by vearutop** [comment, marked as answer]
Named types can express custom schema with help of special interfaces, `Enum` is one of such interfaces. 
Other are listed here: https://github.com/swaggest/jsonschema-go/#implementing-interfaces-on-a-type.

Most important are `Preparer`, `Exposer` and `RawExposer`.

So for this might be a good solution to define `Preparer` that sets all needed fields at once:
```go
func (s SortDir) PrepareJSONSchema(schema *jsonschema.Schema) error {
	schema.Enum = []interface{}{"asc", "desc"}
	schema.WithDefault("asc")
	schema.WithExamples("desc")

	return nil
}
```
`Enum` can be combined with `Preparer` if that brings convenience.

In cases when direct import dependency on `"github.com/swaggest/jsonschema-go"` is unwanted (e.g. in a 3rd-party lib with), it is possible to express schema with lib-agnostic interface:

```go
func (s SortDir) JSONSchemaBytes() ([]byte, error) {
	return []byte(`{"type":"string","enum":["asc","desc"],"default":"asc","examples":["desc"]}`), nil
}
```

Setting up schema with field tags is usually the best way to describe properties, but is less optimal for type-based descriptions (as it can lead to code duplication if same type needs to be described in multiple properties).

**Reply by pboguslawski** [reply]
Works fine thank you. Just noticed that default value is not applied for input (see https://github.com/swaggest/rest/issues/133).
